### PR TITLE
RAC-437: Add Input & Field components

### PR DESCRIPTION
**Pull request by ValentinMumble**
**Description (for Contributor and Core Developer)**

Don't be afraid it's 1800+ lines but 1000+ are for the locale flags 😉 

- Add TextInput component
- Add Field component
- Add useId utility hook
- TODO: add little flag of the locale, that will come in another PR https://github.com/akeneo/pim-community-dev/pull/13530/

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Todo
| Added legacy Behats               | Todo
| Added acceptance tests            | Todo
| Added integration tests           | Todo
| Changelog updated                 | Todo
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

**Review by juliensnz — APPROVED** [review]

File: akeneo-design-system/src/components/Field/Field.stories.mdx [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+  component={Field}
+  argTypes={{
+    children: {table: {type: {summary: 'Input | Helper[]'}}},
+  }}
+  args={{
+    label: 'My field label',
+  }}
+/>
+
+# Field
```
Comment (StevenVAIDIE) [inline akeneo-design-system/src/components/Field/Field.stories.mdx:18]:
```suggestion
# Field (Beta)
```

File: akeneo-design-system/src/components/Input/TextInput/TextInput.stories.mdx [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+<Meta
+  title="Components/Inputs/TextInput"
+  component={TextInput}
+  args={{
+    value: '',
+    placeholder: 'Please enter a value in the TextField',
+  }}
+/>
+
+# TextInput
```
Comment (StevenVAIDIE) [inline akeneo-design-system/src/components/Input/TextInput/TextInput.stories.mdx:14]:
```suggestion
# TextInput (Beta)
```

File: akeneo-design-system/src/components/Input/TextInput/TextInput.tsx [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    onChange?: (newValue: string) => void;
+  }
+>;
+
+/**
+ * The TextInput component allows the user to enter content and data when the expected input is a single line of text.
+ */
+const TextInput = React.forwardRef<HTMLInputElement, TextInputProps>(
+  ({invalid, onChange, readOnly, characterLeftLabel, ...rest}: TextInputProps, forwardedRef: Ref<HTMLInputElement>) => {
+    const handleChange = useCallback(
```
Comment (StevenVAIDIE) [inline akeneo-design-system/src/components/Input/TextInput/TextInput.tsx:90]: Should we throw an Error when input is not read only and there is no onChange ? Maybe we can enforce it by the same way of BooleanProps (https://github.com/akeneo/pim-community-dev/pull/13525/files?short_path=0c9323d#diff-ea18ef73ba4f8ee3df3744190e9dd750bdaf3b7c4214fd2f660b724b39caf946R102)

File: akeneo-design-system/src/components/Field/Field.stories.mdx [hunk 4]
```diff
@@ -0,0 +1,106 @@
+import {useState} from 'react';
+import {Meta, Story, ArgsTable, Canvas} from '@storybook/addon-docs/blocks';
+import {Field} from './Field.tsx';
```
Comment (StevenVAIDIE) [inline akeneo-design-system/src/components/Field/Field.stories.mdx:3]: :eyes:

File: akeneo-design-system/src/components/Field/Field.tsx [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+  ({label, locale, channel, incomplete = false, children, ...rest}: FieldProps, forwardedRef: Ref<HTMLDivElement>) => {
+    const id = useId('field_');
+
+    const decoratedChildren = React.Children.map(children, child => {
+      if (React.isValidElement<HelperProps>(child) && child.type === Helper) {
+        return <HelperContainer>{React.cloneElement(child, {inline: true})}</HelperContainer>;
+      }
+
+      if (React.isValidElement<InputProps>(child)) {
+        return React.cloneElement(child, {id});
```
[the remark this answers is not in the document]
Comment (StevenVAIDIE) [reply akeneo-design-system/src/components/Field/Field.tsx:89]: add aria-labelledby ?

**Review by StevenVAIDIE — APPROVED** [review]
Can you delete all old boolean screenshots ?

**Comment by ValentinMumble** [comment]
> Can you delete all old boolean screenshots ?

@StevenVAIDIE it's done https://github.com/akeneo/pim-community-dev/pull/13525/files#diff-93c2c104675567cdbe330d808ebd043942674328645136d8d90f3d60e0efcac6

File: akeneo-design-system/src/components/Field/Field.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (juliensnz) [reply akeneo-design-system/src/components/Field/Field.tsx:89]: If there is multiple "html input" in a DSM input (for example price collection) the DSM input will expose only one component to the outside and will be responsible of assigning the id to the right html input (the first one for example). What do you think about this proposition?

File: akeneo-design-system/src/components/Field/Field.unit.tsx [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+      <Helper>Some info</Helper>
+      <Helper level="error">Another one</Helper>
+    </Field>
+  );
+
+  expect(screen.getByTestId('text-input')).toBeInTheDocument();
+  expect(screen.getByText('Some info')).toBeInTheDocument();
+  expect(screen.getByText('Another one')).toBeInTheDocument();
+  expect(screen.queryByText('Something else')).not.toBeInTheDocument();
+});
```
[the remark this answers is not in the document]
Comment (juliensnz) [reply akeneo-design-system/src/components/Field/Field.unit.tsx:33]: As we are doing a PR on the side, we didn't do it right away, but you are right, we should add it :)

File: akeneo-design-system/src/components/Field/Field.stories.mdx (hunk 4 above)
Comment (ValentinMumble) [reply akeneo-design-system/src/components/Field/Field.stories.mdx:3]: 👀

File: akeneo-design-system/src/components/Input/TextInput/TextInput.tsx (hunk 3 above)
Comment (ValentinMumble) [reply akeneo-design-system/src/components/Input/TextInput/TextInput.tsx:88]: Enforced by type here https://github.com/akeneo/pim-community-dev/pull/13525/commits/0f4a0bdd1cf30dde04bebbf7b4732a367c05f183

File: akeneo-design-system/src/components/Field/Field.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (StevenVAIDIE) [reply akeneo-design-system/src/components/Field/Field.tsx]: Yes it's good for me

**Pull request merged**